allennlp/data/fields/production_rule_field.py

```diff
@@ -0,0 +1,196 @@
+from typing import Any, Dict, List, Set, Tuple
+
+import torch
+from torch.autograd import Variable
+from overrides import overrides
+
+from allennlp.data.fields.field import Field
+from allennlp.data.tokenizers.token import Token
+from allennlp.data.token_indexers.token_indexer import TokenIndexer
+from allennlp.data.vocabulary import Vocabulary
+
+ProductionRuleArray = Dict[str, Tuple[str, bool, Dict[str, torch.Tensor]]]  # pylint: disable=invalid-name
+
+# mypy doesn't like that we're using a crazy data type - the data type we use here is _supposed_ to
+# be in the bounds of DataArray, but ProductionRuleArray definitely isn't.  TODO(mattg): maybe we
+# should find a better way to loosen those bounds, or let people extend them.  E.g., we could have
+# DataArray be a class, and let people subclass it, or something.
+class ProductionRuleField(Field[ProductionRuleArray]):  # type: ignore
+    """
+    This ``Field`` represents a production rule from a grammar, like "S -> [NP, VP]", "N -> John",
+    or "<b,c> -> [<a,<b,c>>, a]".
+
+    We assume a few things about how these rules are formatted:
+
+        - There is a left-hand side (LHS) and a right-hand side (RHS), where the LHS is always a
+          non-terminal, and the RHS is either a terminal, a non-terminal, or a sequence of
+          non-terminals (for now, you can't mix terminals and non-terminals in the RHS).
```

Comment:
Okay, I see that for the purpose of this class it is fine to consider `lambda x` a "non-terminal", because the only difference between "terminals" and "non-terminals", as you call them in this class is how you index them, and you won't need to index `lambda x` differently from other non-terminals.
But then, the distinction isn't really between terminals and non-terminals. It is between table (or world) specific items, and general items. For example, I don't see a need for indexing `<r,d>` and `argmax` differently, but you would want to do something different for `fb:cell.medals`. But then again, you can define the indexer in such a way as to look at the terminal and do the right thing. If that is what you have in mind, you may want to change the docstring here to make that clear. The docstring is great otherwise.

Reply:
The main issue is in how the RHS is treated - if we see "[...]", we use the non-terminal indexer, which typically is just a vocab lookup.  If you have arbitrary terminal strings inside a RHS sequence, you'll get unknown actions at test time.  You could imagine trying to index each item in that list separately, and then letting the model decide how to combine them, but that gets really complicated really quickly.

And yes, if you have terminals in the grammar that you want to be indexed as if they were non-terminals, you could just include them in the set of "nonterminals" that's passed in the constructor.  I'll make a note of that in the docstring.